l1-contracts/src/core/messagebridge/NewOutbox.sol

```diff
@@ -0,0 +1,116 @@
+
+/**
+ * @title NewOutbox
+ * @author Aztec Labs
+ * @notice Lives on L1 and is used to consume L2 -> L1 messages. Messages are inserted by the state transitioner
+ * and will be consumed by the portal contracts.
+ */
+contract NewOutbox is INewOutbox {
+  using Hash for DataStructures.L2ToL1Msg;
+
+  struct RootData {
+    bytes32 root;
+    uint256 height;
+    mapping(uint256 => bool) nullified;
+  }
+
+  address public immutable STATE_TRANSITIONER;
+  mapping(uint256 l2BlockNumber => RootData) public roots;
+
+  constructor(address _stateTransitioner) {
+    STATE_TRANSITIONER = _stateTransitioner;
+  }
+
+  /*
+   * @notice Inserts the root of a merkle tree containing all of the L2 to L1 messages in
+   * a block specified by _l2BlockNumber.
+   * @dev Only callable by the state transitioner (rollup contract)
+   * @dev Emits `RootAdded` upon inserting the root successfully
+   * @param _l2BlockNumber - The L2 Block Number in which the L2 to L1 messages reside
+   * @param _root - The merkle root of the tree where all the L2 to L1 messages are leaves
+   * @param _height - The height of the merkle tree that the root corresponds to
+   */
+  function insert(uint256 _l2BlockNumber, bytes32 _root, uint256 _height)
+    external
+    override(INewOutbox)
+  {
+    if (msg.sender != STATE_TRANSITIONER) {
+      revert Errors.Outbox__Unauthorized();
+    }
+
+    if (roots[_l2BlockNumber].root != bytes32(0)) {
+      revert Errors.Outbox__RootAlreadySetAtBlock(_l2BlockNumber);
+    }
+
+    if (_root == bytes32(0)) {
+      revert Errors.Outbox__InsertingInvalidRoot();
+    }
+
+    roots[_l2BlockNumber].root = _root;
+    roots[_l2BlockNumber].height = _height;
+
+    emit RootAdded(_l2BlockNumber, _root, _height);
+  }
+
+  /*
+   * @notice Consumes an entry from the Outbox
+   * @dev Only useable by portals / recipients of messages
+   * @dev Emits `MessageConsumed` when consuming messages
+   * @param _l2BlockNumber - The block number specifying the block that contains the message we want to consume
+   * @param _leafIndex - The index inside the merkle tree where the message is located
+   * @param _message - The L2 to L1 message
+   * @param _path - The sibling path used to prove inclusion of the message, the _path length directly depends
+   * on the total amount of L2 to L1 messages in the block. i.e. the length of _path is equal to the depth of the
+   * L1 to L2 message tree.
+   */
+  function consume(
+    uint256 _l2BlockNumber,
+    uint256 _leafIndex,
+    DataStructures.L2ToL1Msg memory _message,
+    bytes32[] memory _path
+  ) external override(INewOutbox) {
+    if (msg.sender != _message.recipient.actor) {
+      revert Errors.Outbox__InvalidRecipient(_message.recipient.actor, msg.sender);
+    }
+
+    if (block.chainid != _message.recipient.chainId) {
+      revert Errors.Outbox__InvalidChainId();
+    }
+
+    RootData storage rootData = roots[_l2BlockNumber];
+
+    bytes32 expectedRoot = rootData.root;
```

Comment:
I'd rename this to e.g. block root. The fact that it's 'expected' is more of a verfication concern, I found it odd to think about verification so early in the lifecycle here.

Reply:
`blockRoot` and `treeHeight` are fine I think. Height is used in a single place, so it's meaning is quite clear.